metadata/src/main/java/org/apache/kafka/queue/EventQueue.java

```diff
@@ -0,0 +1,263 @@
+            return OptionalLong.of(deadlineNs);
+        }
+    }
+
+    class EarliestDeadlineFunction implements Function<OptionalLong, OptionalLong> {
+        private final long newDeadlineNs;
+
+        public EarliestDeadlineFunction(long newDeadlineNs) {
+            this.newDeadlineNs = newDeadlineNs;
+        }
+
+        @Override
+        public OptionalLong apply(OptionalLong prevDeadlineNs) {
+            if (!prevDeadlineNs.isPresent()) {
+                return OptionalLong.of(newDeadlineNs);
+            } else if (prevDeadlineNs.getAsLong() < newDeadlineNs) {
+                return prevDeadlineNs;
+            } else {
+                return OptionalLong.of(newDeadlineNs);
+            }
+        }
+    }
+
+    class VoidEvent implements Event {
+        public final static VoidEvent INSTANCE = new VoidEvent();
+
+        @Override
+        public void run() throws Exception {
+        }
+    }
+
+    /**
+     * Add an element to the front of the queue.
+     *
+     * @param event             The mandatory event to prepend.
+     */
+    default void prepend(Event event) {
+        enqueue(EventInsertionType.PREPEND, null, NoDeadlineFunction.INSTANCE, event);
+    }
+
+    /**
+     * Add an element to the end of the queue.
+     *
+     * @param event             The event to append.
+     */
+    default void append(Event event) {
+        enqueue(EventInsertionType.APPEND, null, NoDeadlineFunction.INSTANCE, event);
+    }
+
+    /**
+     * Add an event to the end of the queue.
+     *
+     * @param deadlineNs        The deadline for starting the event, in monotonic
+     *                          nanoseconds.  If the event has not started by this
+     *                          deadline, handleException is called with a
+     *                          @{org.apache.kafka.common.errors.TimeoutException},
+     *                          and the event is cancelled.
+     * @param event             The event to append.
+     */
+    default void appendWithDeadline(long deadlineNs, Event event) {
+        enqueue(EventInsertionType.APPEND, null, new DeadlineFunction(deadlineNs), event);
+    }
+
+    /**
+     * Schedule an event to be run at a specific time.
+     *
+     * @param tag                   If this is non-null, the unique tag to use for this
+     *                              event.  If an event with this tag already exists, it
+     *                              will be cancelled.
+     * @param deadlineNsCalculator  A function which takes as an argument the existing
+     *                              deadline for the event with this tag (or empty if the
+     *                              event has no tag, or if there is none such), and
+     *                              produces the deadline to use for this event.
+     *                              Once the deadline has arrived, the event will be
+     *                              prepended to the queue.  Events whose deadlines are
```

Comment:
"the event will be prepended to the queue" : This is no longer true in the implementation.